shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -1256,31 +1354,38 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
         } else typeTerms(stmts, false, Nil)(ctx.nest, raise, prov, vars, genLambdas)
       case Bind(l, r) =>
         val l_ty = typeMonomorphicTerm(l)
-        val newCtx = ctx.nest // so the pattern's context don't merge with the outer context!
+        val newCtx = if (ctx.inQuote) ctx.enterQuotedScope else ctx.nest // so the pattern's context don't merge with the outer context!
         val r_ty = typePattern(r)(newCtx, raise)
         ctx ++= newCtx.env
+        solveQuoteContext(ctx, newCtx)
         con(l_ty, r_ty, r_ty)
       case Test(l, r) =>
         val l_ty = typeMonomorphicTerm(l)
-        val newCtx = ctx.nest
+        val newCtx = if (ctx.inQuote) ctx.enterQuotedScope else ctx.nest
         val r_ty = typePattern(r)(newCtx, raise) // TODO make these bindings flow
+        solveQuoteContext(ctx, newCtx)
         con(l_ty, r_ty, TopType)
         BoolType
       case With(t, rcd) =>
         val t_ty = typeMonomorphicTerm(t)
         val rcd_ty = typeMonomorphicTerm(rcd)
         (t_ty without rcd.fields.iterator.map(_._1).toSortedSet) & (rcd_ty, prov)
       case CaseOf(s, cs) =>
-        val s_ty = typeMonomorphicTerm(s)
-        if (newDefs) con(s_ty, ObjType.withProv(prov), TopType)
+        val midCtx = if (ctx.inQuote) ctx.enterQuotedScope else ctx
```

Comment:
I try to avoid this as it's quite error-prone. Better do:
```scala
(if (ctx.inQuote) ctx.enterQuotedScope else ctx) |> { implicit ctx => ... }
```
so we can't forget to pass this `midCtx` thing explicitly everywhere.